src/js/stores/document.js

```diff
@@ -166,18 +173,37 @@ define(function (require, exports, module) {
             if (suppressChange) {
                 return;
             }
-
-            // If some selected layer remains uninitialized, a new document will
-            // come along shortly. Wait until then to trigger the change event.
-            var initialized = nextDocument.layers && nextDocument.layers.selected
-                .every(function (layer) {
-                    return layer.initialized;
-                });
-
-            if (initialized) {
-                // Include the document in the event payload
-                this.emit(changeEventName || "change", nextDocument);
+            
+            var promise = Promise.bind(this);
+
+            // Notify change of layer selection state.
+            if (oldDocument) {
+                promise = promise.then(function () {
+                    oldDocument.layers.layers.forEach(function (layer) {
+                        var nextLayer = nextDocument.layers.byID(layer.id),
+                            callback = this._layerIDToStateListenner[layer.id];
+                        
+                        if (nextLayer && callback &&
```

Comment:
If `!nextLayer || !callback` is there a bug? Should we warn or possibly throw in that case? If not, can you add a comment about why it might be expected for one of these to not hold?